File: AddOns/Share/TKShareHelper.m
//
//  ShareHelper.m
//  TripGo
//
//  Created by Adrian Schoenig on 15/11/2013.
//
//

#import "TKShareHelper.h"

#import <TripKit/TKTripKit.h>

@import SGSearchKit;

@implementation TKShareHelper

#pragma mark - Meet URL

+ (BOOL)isMeetURL:(NSURL *)url
{
  return [[url path] isEqualToString:@"/meet"];
}

+ (NSURL *)meetURLForCoordinate:(CLLocationCoordinate2D)coordinate
                         atTime:(NSDate *)time
{
  return [self meetURLForCoordinate:coordinate atTime:time baseURL:@"http://tripgo.me"];
}

+ (NSURL *)meetURLForCoordinate:(CLLocationCoordinate2D)coordinate
                         atTime:(NSDate *)time
                        baseURL:(NSString *)baseURL
{
  NSString *urlString = [NSString stringWithFormat:@"%@/meet?lat=%.5f&lng=%.5f&at=%.0f", baseURL, coordinate.latitude, coordinate.longitude, [time timeIntervalSince1970]];
  return [NSURL URLWithString:urlString];
}

#pragma mark - Query URL

+ (BOOL)isQueryURL:(NSURL *)url
{
  return [[url path] isEqualToString:@"/go"];
}

+ (NSURL *)queryURLForStart:(CLLocationCoordinate2D)start
                        end:(CLLocationCoordinate2D)end
                   timeType:(SGTimeType)timeType
                       time:(nullable NSDate *)time
{
  return [self queryURLForStart:start end:end timeType:timeType time:time baseURL:@"http://tripgo.me"];
}

+ (NSURL *)queryURLForStart:(CLLocationCoordinate2D)start
                        end:(CLLocationCoordinate2D)end
                   timeType:(SGTimeType)timeType
                       time:(nullable NSDate *)time
                    baseURL:(NSString *)baseURL
{
  NSMutableString *urlString = [NSMutableString stringWithFormat:@"%@/go?tlat=%.5f&tlng=%.5f", baseURL, end.latitude, end.longitude];
  if (CLLocationCoordinate2DIsValid(start)) {
    [urlString appendFormat:@"&flat=%.5f&flng=%.5f", start.latitude, start.longitude];
  }
  if (time && timeType != SGTimeTypeLeaveASAP) {
    [urlString appendFormat:@"&time=%.0f&type=%ld", [time timeIntervalSince1970], (long)timeType];
  }
  return [NSURL URLWithString:urlString];
}

+ (void)geocodeString:(NSString *)string
        usingGeocoder:(id<SGGeocoder>)geocoder
           completion:(void(^)( SGKNamedCoordinate * _Nullable coordinate))completion
{
  [geocoder geocodeString:string
               nearRegion:MKMapRectWorld
                  success:
   ^(NSString * _Nonnull query, NSArray<SGKNamedCoordinate *> * _Nonnull results) {
#pragma unused(query)
    dispatch_async(dispatch_get_main_queue(), ^{
      id<MKAnnotation> annotation = [SGBaseGeocoder pickBestFromResults:results];
      if (annotation) {
        SGKNamedCoordinate *coordinate = [SGKNamedCoordinate namedCoordinateForAnnotation:annotation];
        coordinate.name = string;
        completion(coordinate);
      } else {
        completion(nil);
      }
    });
  } failure:
   ^(NSString * _Nonnull query, NSError * _Nullable error) {
#pragma unused(query, error)
     completion(nil);
  }];
  
}

#pragma mark - Stops

+ (BOOL)isStopURL:(NSURL *)url {
  return [[url path] containsString:@"/stop"];
}

+ (NSURL *)stopURLForStopCode:(NSString *)stopCode
                inRegionNamed:(NSString *)regionName
                       filter:(NSString *)filter
{
  return [self stopURLForStopCode:stopCode inRegionNamed:regionName filter:filter baseURL:@"http://tripgo.me"];
}

+ (NSURL *)stopURLForStopCode:(NSString *)stopCode
                inRegionNamed:(NSString *)regionName
                       filter:(NSString *)filter
                      baseURL:(NSString *)baseURL
{
  NSString *addendum;
  if (filter) {
    addendum = [filter stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
  } else {
    addendum = @"";
  }
  
  NSString *escapedStopCode = [stopCode stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
  NSString *urlString = [NSString stringWithFormat:@"%@/stop/%@/%@/%@", baseURL, regionName, escapedStopCode, addendum];
  return [NSURL URLWithString:urlString];
}

#pragma mark - Services

+ (BOOL)isServicesURL:(NSURL *)url {
  return [[url path] isEqualToString:@"/service"];
}

+ (NSURL *)serviceURLForServiceID:(NSString *)serviceID
                       atStopCode:(NSString *)stopCode
                    inRegionNamed:(NSString *)regionName
{
  return [self serviceURLForServiceID:serviceID atStopCode:stopCode inRegionNamed:regionName baseURL:@"http://tripgo.me"];
}

+ (NSURL *)serviceURLForServiceID:(NSString *)serviceID
                       atStopCode:(NSString *)stopCode
                    inRegionNamed:(NSString *)regionName
                          baseURL:(NSString *)baseURL
{
  NSString *escapedStopCode = [stopCode stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
  NSString *escapedServiceID = [serviceID stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
  NSString *urlString = [NSString stringWithFormat:@"%@/service?regionName=%@&stopCode=%@&serviceID=%@", baseURL, regionName, escapedStopCode, escapedServiceID];
  return [NSURL URLWithString:urlString];
}

+ (void)serviceDetailsForURL:(NSURL *)url
                     details:(void (^)(NSString *stopCode, NSString *regionName, NSString *serviceID))detailBlock {
  // re-construct the parameters
  NSArray *queryComponents = [[url query] componentsSeparatedByString:@"&"];
  NSMutableDictionary *params = [NSMutableDictionary dictionaryWithCapacity:queryComponents.count];
  for (NSString *param in queryComponents) {
    NSArray *elements = [param componentsSeparatedByString:@"="];
    if (elements.count == 2) {
      params[elements[0]] = elements[1];
    }
  }
  
  // construct the request
  if (! params[@"stopCode"] || ! params[@"regionName"] || ! params[@"serviceID"])
    return;
  
  NSString *regionName = params[@"regionName"];
  NSString *stopCode = [params[@"stopCode"] stringByRemovingPercentEncoding];
  NSString *servideId = [params[@"serviceID"] stringByRemovingPercentEncoding];
  
  detailBlock(stopCode, regionName, servideId);
}

@end

File: TripKit/AddOns/Share/TKShareHelper.swift
//
//  TKShareHelper.swift
//  TripKit
//
//  Created by Adrian Schoenig on 30/6/17.
//  Copyright © 2017 SkedGo. All rights reserved.
//

import Foundation
import CoreLocation

public class TKShareHelper: NSObject {
  private override init() {
    super.init()
  }
}

// MARK: - Query URLs

//+ (void)geocodeString:(NSString *)string
//usingGeocoder:(id<SGGeocoder>)geocoder
//completion:(void(^)( SGKNamedCoordinate * _Nullable coordinate))completion
//{
//  [geocoder geocodeString:string
//    nearRegion:MKMapRectWorld
//    success:
//    ^(NSString * _Nonnull query, NSArray<SGKNamedCoordinate *> * _Nonnull results) {
//    #pragma unused(query)
//    dispatch_async(dispatch_get_main_queue(), ^{
//    id<MKAnnotation> annotation = [SGBaseGeocoder pickBestFromResults:results];
//    if (annotation) {
//    SGKNamedCoordinate *coordinate = [SGKNamedCoordinate namedCoordinateForAnnotation:annotation];
//    coordinate.name = string;
//    completion(coordinate);
//    } else {
//    completion(nil);
//    }
//    });
//    } failure:
//    ^(NSString * _Nonnull query, NSError * _Nullable error) {
//    #pragma unused(query, error)
//    completion(nil);
//    }];
//  
//}


extension TKShareHelper {
  
  @objc public static func isQueryURL(_ url: URL) -> Bool {
    return url.path.hasPrefix("/go")
  }

  @objc public static func createQueryURL(start: CLLocationCoordinate2D, end: CLLocationCoordinate2D, timeType: SGTimeType, time: Date?) -> URL {
    return createQueryURL(start: start, end: end, timeType: timeType, time: time, baseURL: "https://tripgo.com")
  }
  
  
  @objc public static func createQueryURL(start: CLLocationCoordinate2D, end: CLLocationCoordinate2D, timeType: SGTimeType, time: Date?, baseURL: String) -> URL {
    
    // TODO: use format string and truncate lat/lng after 5 decimals
    var urlString = "\(baseURL)/go?tlat=\(end.latitude)&tlng=\(end.longitude)"
    if start.isValid {
      urlString.append("&flat=\(start.latitude)&flng=\(start.longitude)")
    }
    
    if let time = time, timeType != .leaveASAP {
      urlString.append("&time=\(Int(time.timeIntervalSince1970))&type=\(timeType.rawValue)")
    }
    
    return URL(string: urlString)!
  }
  
  
}


// MARK: - Meet URLs

extension TKShareHelper {

  @objc public static func isMeetURL(_ url: URL) -> Bool {
    return url.path.hasPrefix("/meet")
  }
  
  @objc public static func createMeetURL(coordinate: CLLocationCoordinate2D, at time: Date, baseURL: String = "https://tripgo.com") -> URL {
    let urlString = "\(baseURL)/meet?lat=\(coordinate.latitude)&lng=\(coordinate.longitude)&at=\(Int(time.timeIntervalSince1970))"
    return URL(string: urlString)!
  }
  
}

// MARK: - Stop URLs

extension TKShareHelper {

  @objc public static func isStopURL(_ url: URL) -> Bool {
    return url.path.hasPrefix("/stop")
  }

  @objc public static func createStopURL(stopCode: String, inRegionNamed regionName: String, filter: String?) -> URL {
    return createStopURL(stopCode: stopCode, inRegionNamed: regionName, filter: filter, baseURL: "https://tripgo.com")
  }
  
  @objc public static func createStopURL(stopCode: String, inRegionNamed regionName: String, filter: String?, baseURL: String) -> URL {

    let escapedCode = (stopCode as NSString).addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)!
    
    let addendum: String
    if let filter = filter, let escaped = (filter as NSString).addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
      addendum = escaped
    } else {
      addendum = ""
    }
    
    let urlString = "\(baseURL)/stop/\(regionName)/\(escapedCode)/\(addendum)"
    return URL(string: urlString)!
  }
  
}

// MARK: - Service URLs

extension TKShareHelper {
  
  @objc
  public static func isServiceURL(_ url: URL) -> Bool {
    return url.path.hasPrefix("/service")
  }
  
  @objc
  public static func createServiceURL(serviceID: String, atStopCode stopCode: String, inRegionNamed regionName: String, baseURL: String = "https://tripgo.com") -> URL {

    let escapedID = (serviceID as NSString).addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)!
    
    let escapedCode = (stopCode as NSString).addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)!
  
    let urlString = "\(baseURL)/service/\(regionName)/\(escapedCode)/\(escapedID)"
    return URL(string: urlString)!
  }

}
